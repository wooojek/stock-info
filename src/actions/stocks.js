export const addStock = (stock) => ({
  type: 'ADD_STOCK',
  stock,
});

export const setStocks = (stocks) => ({
  type: 'SET_STOCKS',
  stocks,
});

export const startSetStocks = () => {
  return (dispatch, getState) => {
    return new Promise((resolve) => {
      const stocks = JSON.parse(localStorage.getItem('stocks')) || [];
      dispatch(setStocks(stocks));
      resolve();
    });
  };
}