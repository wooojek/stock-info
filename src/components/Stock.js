import React from 'react';
import numeral from 'numeral';

const { ALPHAVANTAGE_API_KEY } = process.env;

export class Stock extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.name,
      symbol: props.symbol,
      market: props.market,
      domain: props.domain || '',
      logo: props.logo || 'https://via.placeholder.com/64',
    }
  }

  getImage = () => {
    const query = this.state.name.split(' ')[0];
    return fetch(`https://autocomplete.clearbit.com/v1/companies/suggest?query=${query}`)
      .then((response) => {
        if (!response.ok) {
          throw Error(response.statusText);
        }
        return response.json();
      })
      .then(([{ domain, logo }]) => {
        return {
          domain,
          logo,
        };
      })
      .catch((e) => console.log(e));
  }

  getStockInfo = () => {
    return fetch(`https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=${this.state.symbol}&apikey=${ALPHAVANTAGE_API_KEY}`)
        .then((response) => {
          if (!response.ok) {
            throw Error(response.statusText);
          }
          return response.json();
        })
        .then((data) => {
          if (data['Note']) {
            throw Error(data['Note']);
          }
          const stockData = data['Global Quote'];
          return {
            price: stockData['05. price'],
            closed: stockData['07. latest trading day'],
            change: stockData['09. change'],
            changePercent: stockData['10. change percent'],
          };
        })
        .catch((e) => console.log(e));
  }

  async componentDidMount() {
    try {
      const { domain, logo } = await this.getImage();
      const details = await this.getStockInfo();

      this.setState(() => ({
        details,
        domain,
        logo,
      }));
    } catch (e) {
      console.log(e);
    }
  }

  render() {
    const details = this.state.details;
    const market = this.state.market;
    return (
      <div className="list-item">
        <div className="stock-container">
          <div
            className="stock-img"
            style={{ backgroundImage: `url(${this.state.logo})` }}
          >
          </div>
          <div className="stock-info">
            <div className="stock-info__names">
              <span className="bold">{this.state.name}</span>
              <span>{this.state.symbol}</span>
              {this.state.domain ? (<span>{this.state.domain}</span>) : false}
            </div>
            {market ? (
              <div className="stock-info__market">
                <span>{market.region}</span>
                <span>{market.marketOpen} - {market.marketClose}</span>
                <span>{market.timezone}</span>
              </div>
            ) : (<div></div>)}
            {details ? (
              <div className="stock-info__details">
                <span className="bold">{numeral(details.price).format('$0,0.00')}</span>
                <span className={details.change >= 0 ? 'stock--up' : 'stock--down'}>
                  {numeral(details.change).format('0.00')} ({numeral(details.changePercent.slice(0, -2)).format('0.00')}%)
                </span>
                <span>Closed: {details.closed}</span>
              </div>
            ) : (<div></div>)}

          </div>
        </div>
      </div>
    )
  }
}

export default Stock;