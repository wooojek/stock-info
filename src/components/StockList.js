import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

export const StockList = ({ stocks }) => (
  <Fragment>
    <h2>Companies</h2>
    <div className="content-container">
      {stocks.length > 0 ?
        (
          <ul className="list-body">
            {stocks.map((stock, index) => <li className="list-item" key={index}>{stock.name}</li>)}
          </ul>
        ) : (
          <div>
            <span>There are no companies yet.</span>
            <Link to="/add">
              <p>Track your first company.</p>
            </Link>
          </div>
        )
      }
    </div>
  </Fragment>
);

const mapStateToProps = (state) => ({
  stocks: state.stocks,
});

export default connect(mapStateToProps)(StockList);