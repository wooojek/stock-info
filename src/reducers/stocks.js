const defaultStockState = [
  {
    name: 'Apple',
  },
  {
    name: 'Google',
  },
];

export default (state = defaultStockState, action) => {
  switch (action.type) {
    case 'ADD_STOCK':
      return [
        action.stock,
        ...state,
      ];
    default:
      return state;
  }
};