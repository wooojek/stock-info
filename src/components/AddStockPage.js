import React from 'react';
import { connect } from 'react-redux';
import StockForm from './StockForm';
import { startAddStock } from '../actions/stocks';

export class AddStockPage extends React.Component {
  onSubmit = (stock) => {
    this.props.startAddStock(stock);
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Add Stock</h2>
          </div>
        </div>
        <div className="content-container">
          <StockForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
};

const mapDispatchToProps = (dispatch) => ({
  startAddStock: (stock) => dispatch(startAddStock(stock)),
});

export default connect(undefined, mapDispatchToProps)(AddStockPage);