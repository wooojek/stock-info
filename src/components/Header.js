import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { logout } from '../actions/auth';

export const Header = ({ logout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/">
          <h1>Stock Info</h1>
        </Link>
        <Link className="header__element--link" to="/add">
          <h3>Add new company</h3>
        </Link>
        <button className="button button--link header__logout" onClick={logout}>Logout</button>
      </div>

    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  logout: () => dispatch(logout()),
});

export default connect(undefined, mapDispatchToProps)(Header);