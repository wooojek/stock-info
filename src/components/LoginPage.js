import React from 'react';
import { connect } from 'react-redux';
import uuid from 'uuid/v1';
import { login } from '../actions/auth';

const uid = uuid();

export const LoginPage = ({ login }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Stock info</h1>
      <p>Be up to date with your stocks!</p>
      <button className="button" onClick={login}>Login</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  login: () => dispatch(login(uid)),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);