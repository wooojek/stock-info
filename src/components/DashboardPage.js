import React from 'react';
import StockList from './StockList';

export const DashboardPage = () => (
  <div className='content-container'>
    <StockList />
  </div>
);

export default DashboardPage;