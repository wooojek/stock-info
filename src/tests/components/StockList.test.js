import React from "react";
import { shallow } from "enzyme";
import { StockList } from '../../components/StockList';
import stocks from '../fixtures/stocks';

test('should render StockList with stocks', () => {
    const wrapper = shallow(<StockList stocks={stocks} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render StockList with empty message', () => {
    const wrapper = shallow(<StockList stocks={[]} />);
    expect(wrapper).toMatchSnapshot();
});