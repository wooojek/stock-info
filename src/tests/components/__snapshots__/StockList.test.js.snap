// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render StockList with empty message 1`] = `
<Fragment>
  <h2>
    Companies
  </h2>
  <div
    className="content-container"
  >
    <div>
      <span>
        There are no companies yet.
      </span>
      <Link
        replace={false}
        to="/add"
      >
        <p>
          Track your first company.
        </p>
      </Link>
    </div>
  </div>
</Fragment>
`;

exports[`should render StockList with stocks 1`] = `
<Fragment>
  <h2>
    Companies
  </h2>
  <div
    className="content-container"
  >
    <ul
      className="list-body"
    >
      <li
        className="list-item"
        key="0"
      >
        Apple
      </li>
      <li
        className="list-item"
        key="1"
      >
        Google
      </li>
    </ul>
  </div>
</Fragment>
`;
